### ISSUE TITLE
Support TM1829

### ISSUE THREAD

**Issue by focalintent** [FastLED/FastLED#8]

**Comment by focalintent** [comment]
Nah.

**Comment by zone31** [comment]
Is it possible to reopen this ticket?
I recently bought a led strip with 50X3 LED's, using the TM1829.
It seems the signal is reversed, compared to other clockless LED controllers.

**Comment by focalintent** [comment]
Yeah - the signal is odd with it - unfortunately, in order to work on properly supporting it, I would need more TM1829 LEDs (my last set got lost about a year ago, and since I never ran into anyone actually sing it I never bothered trying to replace it).  I can put in an order for a set - but not sure when i'll be able to work on it.

**Comment by focalintent** [comment]
Also - do you know if you have the high speed or the low speed version of the chip?

**Comment by zone31** [comment]
Thank you. If you need anything tested please ask.
As i understand, there are only one revision of the chip, which is able work in 800Khz and 1600Khz mode. Mode is defined by the first 8 bits received at startup(Possibly send in 1600Khz mode?).

I was only able to find a crudely translated datasheet, linked [here] (press download PDF)(http://www.datasheet26.com/PDF/775431/TM1829-pdf.html)

**Comment by focalintent** [comment]
Yeah - I have a copy of the data sheet here (a couple copies, actually).  I'll try to get in an order of TM1829 chips this week to do some testing with.

**Comment by zone31** [comment]
That would be great! 
I would like to contribute myself, but i can't seem to get my head around the timing values T1 T2 and T3? in the picture below, can you explain where they fit in? Thanks in advance.
![image](https://cloud.githubusercontent.com/assets/2137304/11197313/fc56b21a-8cbc-11e5-859c-351243b15dbe.png)

**Comment by focalintent** [comment]
What's weird is that the timing in the picture that you just posted doesn't line up with either of the timings that I've got from the data sheet that I have (they're more of a 550khz data rate - 1.8µs per bit vs. 1.2µs per bit for 800khz, or 0.6µs per bit for 1600khz).

That chart that you pasted doesn't even come out of the TM1829 data sheet.

As for how that chart matches to T1, T2, T3 - T1 is the value for T0H, T2 is the value for T1H - T0H, and finally, T3 is the value from (T1H+T1L)-(T1+T2) - roughly.  Basically, the T1 value is how long before we'd go low for a 0, the T2 value is how much longer before we'd go low for a 1, and the T3 value is how long we go before starting the next bit.

What makes the TM1829 a little bit more challenging, however, is that it also inverts the signal.  Instead of High then Low, it's Low then High for each bit.

**Comment by focalintent** [comment]
(The final bit is being careful to make sure that 255 isn't ever sent for the first byte getting written out - otherwise the data gets handled differently)

**Comment by zone31** [comment]
Sorry, wrong image,
![image](https://cloud.githubusercontent.com/assets/2137304/11197639/1aa5f986-8cbf-11e5-852b-fd2405d89408.png)
![image](https://cloud.githubusercontent.com/assets/2137304/11197671/42eacf3e-8cbf-11e5-8124-673383e630d9.png)
I guess i can flip the HI1 to LO1 and vice versa, for testing.

**Comment by focalintent** [comment]
Ok - so the timings that you'll want to start with are going to be roughly: 

NS(300), NS(500), NS(400)

Also  it looks like these want a specific reset sent of holding the line high for 1.2µs -- though I suspect just leaving the line high in between writing out rounds of LED data would be fine.

**Comment by zone31** [comment]
OK, flipped the HI1 and LO1 compiler definitions, and it seems to work somewhat.
I use a arduino uno, so im fiddling with the FMUL values (currently 2,5,3)
The first chip connected to the arduino UNO is always dark blue, but the other ones update as they should. it seems like the color channels are rotated, when i send out R G B i get B G R.

**Comment by focalintent** [comment]
See https://github.com/FastLED/FastLED/wiki/Rgb-calibration to see how to adjust color ordering (since not everything wires their legs RGB internally).

Also it sounds like it may not quite be registering 1's correctly on that first led (but the signal reshaping/writing is working for the rest of the less).  Try 2,6,3 or 2,6,2.

**Comment by KajoVanBuyten** [comment]
Hello, I followed the suggestions here and got FastLED to work with an Arduino UNO and my TM1829 Strips.
But I have the same problem with the first pixel, that is always blue.
In a german forum they had a similar problem with the Neopixel lib. They said that there was a HIGH -> LOW transition at the end of the Code, that was interpreted as Data package by the TM1829. They shifted it to LOW -> HIGH to fix the first pixel.
I just understand a litte of the Code and can't figure it out myself. It would be really nice if you could make a suggestion where to look at.

**Comment by misel228** [comment]
Would it be of any help  if I sent you a left-over part of my RGB-strip?

**Comment by janLo** [comment]
What is the state of the TM1829 support in the library? I found some traces in the code but it seems to be unfinished? What is missing? And does anyone know if the TM1829 works with the support for the 1809?

**Comment by focalintent** [comment]
No - the TM1829 is different enough from the 1809 that it won't work.  There's the beginnings of support in the library but it never got far because of the weirdness around its inverted signal and such.

**Comment by markusrudolf** [comment]
I would also be in favour of having support for the TM1829 in. I have strips for testing, preferrably with ESP8266

**Comment by thymjan** [comment]
Did you know, there is a modified adafruit neopixel library which can handle the TM1829?
Have a look here (it's in german):
https://blog.thesen.eu/aldi-rgb-led-strip-iclite-auf-tms1829-basis-mit-dem-arduino-steuern/
Here is the download link:
https://blog.thesen.eu/wp-content/uploads/2015/01/Aldi_NeoPixel.zip

May this help?

**Comment by h3ndrik** [comment]
My TM1829 works fine with the following timings in _chipset.h_:
`class TM1829Controller800Khz : public ClocklessController<DATA_PIN, NS(300), NS(500), NS(500), RGB_ORDER, 0, true, 500> {};`

But the logic levels have to be inverted. And at least for the esp8266 that seems not to be the case!? Isn't "FLIP" supposed to do that? Because i have to manually exchange hi() and lo() in _clockless_esp8266.h_ also.

**Comment by focalintent** [comment]
Flip isnt supported on all the platforms at the moment.  Also didn’t think anyone actually sold/used the TM1829 - glad to hear there’s some out there.

**Comment by markusrudolf** [comment]
> Also didn’t think anyone actually sold/used the TM1829

A German supermarket chain ("Aldi") sold LED strips with a seperat controller and IR remote. The strips are using TM1829 in groups of three LEDs. I wanted to use an ESP8266 in place of the controller. I dig out the project then :-)

**Comment by mitchsmith512** [comment]
> Also didn’t think anyone actually sold/used the TM1829
I have 3 sets of GE xmas lights that do color-changing/etc per bulb.  The onboard driver/controller is the TM1829.  The original string controllers don't offer much in the way of patterns, so I was looking to program them myself from arduino.

**Comment by h3ndrik** [comment]
Also renkforce "LED Strip Farbwechsel Digital, 5m" distributed by Conrad Electronic in Germany. Uses TM1829

**Comment by doteq** [comment]
In Poland "Obi" sold TM1829 leds. I found this library: https://www.tns-labs.org/iclite-digital-rgb-led-flex-band-aldi-arduino/ It supports TM1829

**Comment by Hitower78** [comment]
HI There
I found nearly 30m of this Stripes in our Company, so i ask me if there are in the libary yet. 
The Modded "ALDI_Neopix" is running fine

**Comment by juleskers** [comment]
@focalintent I have a 5m strip of tm1829 LEDs, with controller (the infamous "Aldi" set sold in Europe), that I'd love to donate to you. Your FastLED library greatly helped me in a project, so I'd like to contribute something back :-)

I'll gladly ship it anywhere you want. Is there a way I can contact privately you for shipping addresses?
(I'd rather not blab out addresses in the open); you can reach me at my firstname.lastname at gmail.com (lets see the spambots figure that out ;-) )

**Comment by zackees** [comment]
Status update after investigating this thread.

**A regression was found and fixed** in #3797 (merged as `bed4df1`).

The inverted-signal problem everyone in this thread hit was, at one point, partially encoded in the controller declaration:

```cpp
class TM1829Controller800Khz : public FASTLED_CLOCKLESS_CONTROLLER<..., RGB_ORDER, 0, true, 500> {};
//                                                                             ^^^^  ^^^
//                                                                             FLIP  WAIT_TIME
```

Commit `fc0c3d5` (Oct 2025) refactored `src/chipsets.h` and swept away trailing `0, false` template defaults. That was a no-op for every other chipset — but TM1829 was the only chipset passing *non-default* values, and `0, true, 500` got stripped along with them. Both the 800 kHz and 1600 kHz variants lost it, and `TX1813N1` (#1362) inherited the regression.

#3797 restores both parameters.

**Keeping this issue open**, because that restores prior behavior rather than completing TM1829 support. The honest state today:

- `WAIT_TIME=500` is honored via `CMinWait<WAIT_TIME>` on ~20 backends (AVR, ATtiny, ARM bit-bang paths) — the 500 µs latch gap is genuinely back there.
- `FLIP=true` is consumed by **exactly one** backend in the whole tree: nRF52 PWM (`clockless_arm_nrf52.h:53-54`). Every other clockless controller accepts `FLIP` as a template parameter and never references it again — so @h3ndrik's 2018 observation that "FLIP doesn't do anything on ESP8266" is still accurate, and applies to nearly every platform.
- The newer `fl::channels` path (used by ESP32 LCD_CAM/UART drivers) has no polarity concept at all — `ChipsetTimingConfig` carries only `t1/t2/t3/reset`.

**For whoever picks this up:** the hardware inversion flags already exist and sit unused, which makes this much less work than it looks:

| Peripheral | Existing flag | Current value |
|---|---|---|
| ESP32 RMT5 | `irmt5_peripheral.h:94` `bool invert_out` | always `false` |
| ESP32 RMT4 | `irmt4_peripheral.h:177` `setGpio(..., bool invert)` | callers pass `false` |
| ESP32 I2S | `ii2s_peripheral_esp32dev.h:107` `u16 mInvertMask` | always `0` |

The missing plumbing is a polarity bool on `ChipsetTimingConfig` / `ClocklessChipset` (plus `operator==` and `makeTimingConfig`), bridged from the `FLIP` template parameter in `clockless_controller_impl.h`, then wired into those three flags.

Also still unaddressed and worth preserving from this thread — @focalintent and @h3ndrik both flagged it: sending `0xFF` as the first byte puts the chip into constant-current mode, so brightness 255 must be clamped. The original 2014 `f3ef3ff` checkpoint had a commented-out `scale8(scale, 254)` shim for exactly this.

Hardware for testing was offered by @juleskers, @misel228, and @Hitower78.

**Comment by zackees** [comment]
Deferred by automated triage: core FLIP regression fixed in #3797; full inverted-signal support across non-nRF52 backends remains open and needs the ChipsetTimingConfig polarity plumbing described in the issue thread.

**Comment by juleskers** [comment]
Oh wow! I didn't expect to ever get pings from this anymore. Such memories! 🧡 
Great to see a revival of FASTled. LLM are making a lot possible.

I've moved on from LED projects, but I just checked and I still have the tm1829 LED strip, complete with power supply, (non-programmable) controller and IR-remote (picture attached, still in original packaging after three moves...).

<img width="4000" height="3000" alt="Image" src="https://github.com/user-attachments/assets/128c2e25-ce5c-4b4e-9c7c-124ed576934e" />

These were sold by the European discount supermarket Aldi in at least Germany and the Netherlands. Per comments above, "Obi" (a home improvement chain) sold them in Poland. Since Obi also has presence in Germany, there are probably thousands of people with this set in Northern Europe.

The offer to ship it to someone with time/energy to tackle this still stands. However, I won't have the energy to do much else due to other projects (and life).

I can parcel-ship the whole thing, or snip of smaller piece and put it in an envelope, whichever is more convenient.

If someone wants to take up the offer, @-mention me here to discuss shipping details,  or stalk my commits on GitHub to discuss it privately via email.

**Comment by zackees** [comment]
You can ship it to me by reaching out to reddit.com/u/zachvorhies and i'll give you my address.

Yes FastLED is going through a revival. It's on an exponential growth curve at this point as far as development goes.

The biggest thing that excites me is compiling the project in a web browser during development and then deploying once to a real board and having it just work.

Might be worth it to see what we've been up to.

**Comment by juleskers** [comment]
While I lurk some corners of Reddit, I don't maintain an account there. I *think* I've successfully stalked your email address, so (AOL)you've got mail(/AOL) (maybe?)

---

I've had a gander at the recent comment history. Impressive turnout, but I can confidently say I won't have the energy nor raw time to keep up with such an AI-powered firehose 🧡 .

I liked the part where you added hardware loopback devices for "autoresearch". Sounds like exactly the kind of thing that LLMs need to avoid hallucinating into oblivion. And hardware-backed source-of-truth is useful for hhumans too!
I find it spectacularly funny and intrigueing (also seen in other projects) that exactly the kind of 'boring' test harnesses and documentation that humans wish they had written for themselves (but don't),  is what is needed to keep LLM on the rails.
LLM seem to be a lot worse at implicit or large-scale context. But then again, they are effectively on-boarding themselves every few seconds, whereas humans generally take months to familiarize themselves when they change jobs 😅 ..

That said, I consider myself an AI sceptic, waiting for the bubble to pop. And while I admire the commoditisation of programming skills, and the (truely spectacular) leverage experts can tease out, I'm unconvinced as of yet that the societal costs are worth it (stripmining of the internet, electricity/water/dram/GPU scarcity, untested copyright implications of GPL training data, etc).
And of course the pure monetary costs (subsidized lossleading) and anti-sovereign big-tech vendor lock-in.

I'll definitely keep an eye on your work, to learn about what is possible, but the test hardware will have to be the limit of my active contribution (see time/energy, above 😅 )

**Comment by zackees** [comment]
Regardless of whether the AI bubble will or will not pop, I'm betting that everyone needs a platform to run their electronics and that it should be free for everyone and not controlled by corporations. They can do whatever they want, we are still free to make awesome led lighting displays.

**Comment by juleskers** [comment]
> I'm betting that everyone needs a platform to run their electronics and that it should be free for everyone and not controlled by corporations.

On that, we are in full agreement! FOSS is essential to humanity as a great equalizer.

---

As for the email address: did my email arrive? I'd still like to send you the hardware 🧡

**Comment by wida5962loksin** [comment]
Thanks for the update. Yes, I received your message. I’d be happy to
share the email address so you can send the hardware.

**Comment by zackees** [comment]
Replied!
